### PR TITLE
internal/vcs/git: add FindNearestCommit to get commit nearest to a point in time

### PR DESCRIPTION
**Pull request by slimsag**
Stacked on top of #18712

For building historical data of search insights, we need a way to locate the commit
in a repository closest to some given point in time so that we can search it using
a `repo:<repo>@<rev>` filter and determine the number of results that existed around
that point in time. This adds the ability for us to determine which commit is most
appropriate for a given time range.

Signed-off-by: Stephen Gutekanst <stephen@sourcegraph.com>

### CODE REVIEW

File: internal/vcs/git/commits.go [hunk 1]
```diff
@@ -320,6 +329,77 @@ func FirstEverCommit(ctx context.Context, repo api.RepoName) (*Commit, error) {
 	return GetCommit(ctx, repo, id, ResolveRevisionOptions{NoEnsureRevision: true})
 }
 
+// FindNearestCommit finds the commit in the given repository revSpec (e.g. `HEAD` or `mybranch`)
+// whose author date most closely matches the target time.
+//
+// Can return a commit very far away if no nearby one exists.
+// Can theoretically return nil, nil if no commits at all are found.
+func FindNearestCommit(ctx context.Context, repoName api.RepoName, revSpec string, target time.Time) (*Commit, error) {
```
Comment (keegancsmith) [inline internal/vcs/git/commits.go:337]: does it make sense for this to live in vcs/git? It doesn't run any git commands, rather just uses vcs/git functions.

**Review by keegancsmith — APPROVED** [review]
Why is it important to look in both directions? I imagine `Reverse` might be a little expensive for git to do.

**Comment by keegancsmith** [comment]
> Why is it important to look in both directions? I imagine `Reverse` might be a little expensive for git to do.

Oh I see, non-reverse might be very far away from the target date.

File: internal/vcs/git/commits.go [hunk 2]
```diff
@@ -71,6 +71,9 @@ type CommitsOptions struct {
 	After  string // include only commits after this date
 	Before string // include only commits before this date
 
+	Reverse   bool // Whether or not commits should be given in reverse order.
+	DateOrder bool // Whether or not commits should be sorted by date.
```
Comment (efritz) [inline internal/vcs/git/commits.go:75]: What's the default?

File: internal/vcs/git/commits.go (hunk 1 above)
Comment (efritz) [reply internal/vcs/git/commits.go:337]: The interface doesn't make that obvious, though.

**Review by efritz — APPROVED** [review]

**Comment by slimsag** [comment]
> Oh I see, non-reverse might be very far away from the target date.

That is completely right.

File: internal/vcs/git/commits.go [hunk 3]
```diff
@@ -320,6 +329,77 @@ func FirstEverCommit(ctx context.Context, repo api.RepoName) (*Commit, error) {
...
+	}
+	// Resolve e.g. the branch we're looking at.
+	branchCommit, err := ResolveRevision(ctx, repoName, revSpec, ResolveRevisionOptions{NoEnsureRevision: true})
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the next closest commit on or after our target time.
+	commitsAfter, err := Commits(ctx, repoName, CommitsOptions{
+		After:     target.Add(-1 * time.Second).Format(time.RFC3339),
```
[the remark this answers is not in the document]
Comment (slimsag) [reply internal/vcs/git/commits.go:349]: Nope. If you specify `N: 1` / `--count 1` in combination with `--reverse` then you get _the most recent commit_ and not the commit closest to the `After` date. Git does the truncation before the reversal. See https://stackoverflow.com/questions/30832303/git-log-reverse-and-then-limit-the-results-to-one

File: internal/vcs/git/commits.go (hunk 1 above)
Comment (slimsag) [reply internal/vcs/git/commits.go:337]: It's a good question.

Originally I had this in the insights codebase instead, for exactly the reason you said - but I thought there were two benefits to having it here:

1. I didn't have to duplicate the testing functionality we have in this package.
2. If we ever want to switch to libgit, anyone else wants to use this function, someone comes up with a way to optimize it more, etc. ten it's obvious to do it here in isolated fashion.

File: internal/vcs/git/commits.go (hunk 2 above)
Comment (slimsag) [inline internal/vcs/git/commits.go:75]:
```suggestion
	DateOrder bool // Whether or not commits should be sorted by date (optional)
```

File: internal/vcs/git/commits.go [hunk 4]
```diff
@@ -71,6 +71,9 @@ type CommitsOptions struct {
 	After  string // include only commits after this date
 	Before string // include only commits before this date
 
+	Reverse   bool // Whether or not commits should be given in reverse order.
```
Comment (slimsag) [inline internal/vcs/git/commits.go:74]:
```suggestion
	Reverse   bool // Whether or not commits should be given in reverse order (optional)
```

File: internal/vcs/git/commits.go (hunk 2 above)
Comment (slimsag) [reply internal/vcs/git/commits.go]: Clarified both are optional (i.e. zero value / false is the default); this is true of all other fields in this struct.

**Pull request merged**